automerge/src/autotxn.rs

```diff
@@ -0,0 +1,417 @@
+#[derive(Debug, Clone)]
+pub struct AutoTxn {
+    doc: Automerge,
+    transaction: Option<TransactionInner>,
+}
+
+impl Default for AutoTxn {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl AutoTxn {
+    pub fn new() -> Self {
+        Self {
+            doc: Automerge::new(),
+            transaction: None,
+        }
+    }
+
+    /// Get the inner document.
+    #[doc(hidden)]
+    pub fn document(&mut self) -> &Automerge {
+        self.ensure_transaction_closed();
+        &self.doc
+    }
+
+    pub fn set_actor(&mut self, actor: ActorId) {
+        self.ensure_transaction_closed();
+        self.doc.set_actor(actor)
+    }
+
+    pub fn get_actor(&mut self) -> ActorId {
+        self.doc.get_actor()
+    }
+
+    pub fn maybe_get_actor(&self) -> Option<ActorId> {
+        self.doc.maybe_get_actor()
+    }
+
+    pub fn new_with_actor_id(actor: ActorId) -> Self {
+        Self {
+            doc: Automerge::new_with_actor_id(actor),
+            transaction: None,
+        }
+    }
+
+    pub fn pending_ops(&self) -> usize {
+        self.transaction
+            .as_ref()
+            .map(|t| t.pending_ops())
+            .unwrap_or(0)
+    }
+
+    fn try_start_transaction(&mut self) {
+        if self.transaction.is_none() {
+            let actor = self.doc.get_actor_index();
+
+            let seq = self.doc.states.entry(actor).or_default().len() as u64 + 1;
+            let mut deps = self.get_heads();
+            if seq > 1 {
+                let last_hash = self.get_hash(actor, seq - 1).unwrap();
+                if !deps.contains(&last_hash) {
+                    deps.push(last_hash);
+                }
+            }
+
+            self.transaction = Some(TransactionInner {
+                actor,
+                seq,
+                start_op: self.doc.max_op + 1,
+                time: 0,
+                message: None,
+                extra_bytes: Default::default(),
+                hash: None,
+                operations: vec![],
+                deps,
+            });
+        }
+    }
+
+    pub fn fork(&mut self) -> Self {
+        self.ensure_transaction_closed();
+        Self {
+            doc: self.doc.fork(),
+            transaction: self.transaction.clone(),
+        }
+    }
+
+    pub fn commit(&mut self) -> Vec<ChangeHash> {
+        // ensure that even no changes triggers a change
+        self.try_start_transaction();
+        self.transaction
+            .take()
+            .map(|tx| tx.commit(&mut self.doc, None, None))
+            .unwrap_or_else(|| self.doc.get_heads())
+    }
+
+    pub fn commit_with(&mut self, options: CommitOptions) -> Vec<ChangeHash> {
+        self.try_start_transaction();
+        self.transaction
+            .take()
+            .map(|tx| tx.commit(&mut self.doc, options.message, options.time))
+            .unwrap_or_else(|| self.doc.get_heads())
+    }
+
+    pub fn ensure_transaction_closed(&mut self) {
```

Comment:
Could this `ensure_transaction_closed` stuff be avoided by making `AutoTxn` an enum over two possible states. I.e.

```rust
enum AutoTxn {
    InTransaction{
        doc: Automerge,
        transaction: TransactionInner,
    },
    Clean(Automerge),
}
```

This way any code which needs access to the `TransactionInner` would first have to transition the state to `InTransaction`.